### DISCUSSION TITLE
How do I use two fonts in one app?

### DISCUSSION THREAD

**Discussion by egordrachenko** [fyne-io/fyne#4285] [Q&A]
Hi!

I am trying to build and app for my CarPC, and what I would like to do is to use multiple fonts across my app, and I cant find a way to do that. From what I've managed to find during my research, I can create a custom theme and specify a custom font there, but only one, because there's only one field for fonts in theme struct. Am I missing something?

For now I use this crutch to make it work:

```
func (*CarpcTheme) Font(s fyne.TextStyle) fyne.Resource {
	if s.Monospace {
		return resourceAuxMonoRegularTtf
	}
	if s.Bold {
		if s.Italic {
			return theme.DefaultTheme().Font(s)
		}
		return theme.DefaultTheme().Font(s)
	}
	if s.Italic {
		return theme.DefaultTheme().Font(s)
	}
	return theme.DefaultTheme().Font(s)
}
```
So I sacrifice default monospace and replace it with custom resource, so I get my custom font when setting Monospace flag.
Im new to Golang and what I've tried to do from my other languages experience is:

```
type CustomFonts struct {
        AuxMonoFont bool
        OtherFont bool
        fyne.TextStyle
}

func (*CarpcTheme) Font(s CustomFonts) fyne.Resource {
	if s.AuxMono {
		if s.AuxMono.Bold {
			... // returns fyne.Resource
		}
		...
	}

	if s.OtherFont {
		... // returns fyne.Resource
	}
}
```

But It doesn't work as Theme constructor doesn't accept Font func with argument of type that is not fyne.TextStyle, because there's no inheritance in Go.
Another way to do It is to assign each font to specific combination of fyne.TextStyle parameters and then write a function that will take desired font name and return fyne.TextStyle with corresponding set of parameters. This allows to bring enough fonts into app, but... come on, that's insane.

Maybe there's a built-in solution I failed to find or a better workaround, maybe some way to extend fyne.TextStyle?
Appeciate any help, thanks.

**Comment by andydotxyz** [comment]
The aim (and current limitation) is that you can have only 1 font for each text style (permutation if 8 I think between normal/bold/italic/monospace).
It's possible this is a duplicate of request #808.

**Reply by egordrachenko** [reply]
>  only 1 font for each text style

I dont get it. You say that I can have one font for one text style, but how do I specify font for a text style?
```
type TextStyle struct {
	Bold      bool // Should text be bold
	Italic    bool // Should text be italic
	Monospace bool // Use the system monospace font instead of regular
	// Since: 2.2
	Symbol bool // Use the system symbol font.
	// Since: 2.1
	TabWidth int // Width of tabs in spaces
}

```
Font is set in theme and there can be only one font in theme, meaning only one font for a whole app.
I want to reach following behavior: I have FontA.ttf and FontB.ttf and I have LabelA and LabelB, and I want LabelA to be rendered in FontA and LabelB rendered in FontB. Is it possible to reach such behavior?

**Reply by andydotxyz** [reply]
The theme defines which font should be used. It does so based on the TextStyle passed into the Font method.
I'm not sure why you are convinced that limits you to 1 font!

Like I said it's one font per style, not one per theme...

**Reply by egordrachenko** [reply]
Please, dont get me wrong, sir. I'm not a native speaker so I might not express correctly. I'm not convinced that it limits me to one font, I just cant get how to use multiple fonts. It's not obvious from documentation, because there's only one possible Font method in Theme, only one Theme cant be applied to App. And there are no fields in TextStyle struct to specify the font to use. Maybe you could give me a hint how to do that?

P. S. Just to exempt misundersanding - Bold and Italic are different styles of same font. Roboto and Open Sans, for ex., are different fonts. I need to use Roboto and Open Sans (and it's styles) in same one app but in different widgets, and I cant figure out how to do it.

Thank you for response!

**Reply by andydotxyz** [reply]
> Bold and Italic are different styles of same font. 

They don't have to be - you can have a different font for every style (in some cases you must, like for symbol and monospaced). This is why I am confident in saying you are not limited to 1 font.

If however you want to know if you can use the same font style to represent different fonts, that is not currently possible.

**Comment by junmaqiang** [comment]
this is good idea

**Comment by andydotxyz** [comment]
Maybe you will be interested in https://github.com/fyne-io/fyne/pull/4313
